[Identity] Add resend otp feature in login mfa page

LefterisRentas opened the pull request.

Commit:
[Identity] Add resend otp feature in login mfa page

Commit:
[Identity] Rename canResendOtp to isResendDisabled in MFA Login pages

Commit:
[Identity] MFA Login fix build error

dkarkanas commented on src/Indice.Features.Identity.UI/Pages/Bootstrap5/Mfa.cshtml:
```diff
@@ -35,7 +36,17 @@
+                </button>
+                <button type="submit" class="btn btn-info col-5" id="otp-resend" name="Input.ResendOtp" disabled="@isResendDisabled" value="true">
+                    @Localizer["Resend"]
+                    <div indice-if="!isResendDisabled">
```
καλά έκανες και άκουσες το ΑΙ, αλλά τώρα μπήκαν 2 αρνήσεις!!!

οπότε το διαβάζεις σαν "αν δεν ειναι το resend απενεργοποιημενο" ή κατι τετοιο που ήδη κάηκα :D

προσωπικά προτιμώ την "θετική" πλευρά της λογικής, φάση `resendEnabled`, ώστε αν πεις `if (!resendEnabled)` να διαβαστεί σαν "αν δεν ειναι ενεργοποιημένο το resend"

`ή`

κατ' ευθειαν `<div indice-if="!Model.View.HasError">` ?

Τι λες;

LefterisRentas replied:
Θα προτιμήσω το resendEnabled

Commit:
[Identity] MFA Login change back to resendEnabled

cleftheris commented on src/Indice.Features.Identity.UI/Pages/Bootstrap5/Mfa.cshtml:
```diff
@@ -88,6 +99,8 @@
     }
     <script type="text/javascript" src="~/lib/@@fingerprintjs/fingerprintjs/dist/fp.min.js" csp-nonce></script>
     <script type="text/javascript" src="~/js/mfa.js" csp-nonce></script>
+    <script type="text/javascript" src="~/lib/jquery/dist/jquery.min.js" csp-nonce="true"></script>
```
Remove this. As it is available on each framework via LayoutIdentity.cshtml

cleftheris commented on src/Indice.Features.Identity.UI/Pages/Bootstrap5/Mfa.cshtml:
```diff
@@ -9,6 +9,7 @@
     var title = ViewData["Title"] = Localizer["Multifactor Authentication"].Value;
     var maskedPhoneNumber = !string.IsNullOrWhiteSpace(Model.View.User?.PhoneNumber) ? Regex.Replace(Model.View.User.PhoneNumber, @"\d(?!\d{0,1}$)", "X") : null;
     var hubConnectionUrl = Configuration.GetSection("General").GetValue<string>("HubConnectionUrl");
+    var resendEnabled = Model.View.HasError;
```
Why introduce new parameters inside the view that will need to be copied around the different ui frameworks. Since we got a view model this is not sensible to recreate in every cshtml page and also it is susceptible to errors on behalf of the theme developer.

cleftheris commented on src/Indice.Features.Identity.UI/Pages/Mfa.cs:
```diff
@@ -155,6 +156,21 @@ private async Task<MfaLoginViewModel> BuildMfaLoginViewModelAsync(string? return
     }
+
+    private async Task<TotpResult> SendOtp() {
+        var totpService = TotpServiceFactory.Create<User>();
```
Consider throttling/rate limiting. This essentially means that totp service will not send a totp and will fail if the countdown has not passed.

LefterisRentas replied:
This is already being done by the totp service. When pressing the resend button you will see the error from the totp service. The error is not propagated OnGetAsync but only OnPostAsync (resend) we should consider maybe showing the error OnGetAsync too.

cleftheris replied:
I agree

cleftheris commented on src/Indice.Features.Identity.UI/Pages/Tailwind/Mfa.cshtml:
```diff
@@ -9,6 +9,7 @@
     var title = ViewData["Title"] = Localizer["Multifactor Authentication"].Value;
     var maskedPhoneNumber = !string.IsNullOrWhiteSpace(Model.View.User?.PhoneNumber) ? Regex.Replace(Model.View.User.PhoneNumber, @"\d(?!\d{0,1}$)", "X") : null;
     var hubConnectionUrl = Configuration.GetSection("General").GetValue<string>("HubConnectionUrl");
+    var resendEnabled = Model.View.HasError;
```
Same as with previous comment move to view model.

cleftheris reviewed (changes requested):
In general we will need to have similar way of doing things for the addPhone/verifyhpone variants on `MfaOnBoarding` & `ExtendedValidation` flows

cleftheris committed:
Enhance ProfileInputModel documentation and properties

Updated the `ProfileInputModel` class to provide more descriptive summaries for its properties, improving code readability. Introduced a new property `DeveloperTotp` with an accompanying remark about its usage. Minor formatting adjustments were also made for consistency.

cleftheris committed:
[identity] mfa view model enhancements. Try to minimize the custom code needed inside the view itself

Commit:
[Identity]

Refactor: simplify model access and remove jQuery dependency since it was already being provided by the identity layout

Replaced intermediate variables (`maskedPhoneNumber`,
`hubConnectionUrl`, `resendEnabled`) with direct access to
`Model.View` properties for improved clarity and maintainability.

cleftheris committed:
totp result improvements

cleftheris committed:
Merge branch 'feature/identity-resend-otp' of https://github.com/indice-co/Indice.Platform into feature/identity-resend-otp

cleftheris committed:
minor

cleftheris committed:
Add WebHost property to IndiceWebApplicationBuilder

Updated the IndiceWebApplicationBuilder class to include a new WebHost property, allowing direct access to the InnerBuilder's WebHost for enhanced configuration capabilities.

Commit:
[Identity] Refactor MFA pages and update dependencies

Updated NuGet package versions in `Indice.IdentityPages.csproj` to `8.0.0-rc28` and incremented `<VersionSuffix>` in `Directory.Build.props` to `rc29`. Refactored `MfaOnboarding.cshtml`, `MfaOnboardingAddPhone.cshtml`, and `MfaOnboardingverifyPhone.cshtml` to adopt a modernized design using Tailwind CSS. Changes include improved layout with `content-wrapper` and `card-wrapper`, enhanced accessibility, and consistent styling across pages. Focused on improving user experience and responsiveness.

Commit:
Merge branch 'feature/identity-resend-otp' of https://github.com/indice-co/Indice.AspNet into feature/identity-resend-otp

cleftheris reviewed (approved):
Now its ok
